### DISCUSSION TITLE
Yale user-code enable functionality

### DISCUSSION THREAD

**Discussion by mohanblaze** [zwave-js/zwave-js#6717] [Request Support / Investigate Issue] [answered]
### Checklist

- [X] I have read and followed the above instructions
- [X] I have checked the [troubleshooting section](https://zwave-js.github.io/node-zwave-js/#/troubleshooting/index) and my problem is **not** described there.

- [X] I have read the [changelog](https://github.com/zwave-js/node-zwave-js/blob/master/CHANGELOG.md) and my problem was **not** mentioned there or the fix **did not work**.

### Describe the issue

Hi,
We have paired a Yale lock(which supports 250 user codes) and we were able to set the code. The issue is when we add the device, the user_code_v1 CC has an empty list. I mean the user_codes that were set previously are not showing after repairing. We are trying to get the values using the API: **{"api": "writeValue", "args":[{"nodeId":28, "commandClass":99, "endpoint":0, "property": "userIdStatus", "propertyKey":1},1, null]}** but gives a failure response.

We expect it to send the user_code that was set to the device(if already set) when using the above API.

  1. Pair the Yale lock, update the user_code to a few slots, and confirm the codes by controlling(eg: slot1-1111, slot2-2222, slot3-3333, slot4-4444).
  2. Unpair the Yale lock that was paired to the network.
  3. Repair the same Yale lock to another gateway and test it with user codes 1111, 2222, 3333, 4444. These user codes will work since these were set to the device.
  4. Now open the UserCodeV1 section, and you'll find all the user codes with an empty list. Here, I expect the list of user codes that were working for the device(which were set previously).

### Software versions

Driver (node-zwave-js): 11.10.0

### Device information

Manufacturer: Yale Lock
Model name: YRD226
Node ID: 30
manufacturer ID, Product Id, ProductType Id : 297-1536-32770 (0x0129-0x8002-0x0600) 

### Checklist

- [X] I [made sure](https://zwave-js.github.io/node-zwave-js/#/troubleshooting/index?id=rule-1-use-the-driver-log-luke) to provide a **driver log** on level **debug**.

- [X] The log includes a re-interview of the problematic device _(if applicable)_.

- [X] The log includes the problematic interaction with the device _(if applicable)_.

- [X] I provided the node ID of the problematic device _(if applicable)_.

### Upload Logfile

[zwavejs_2024-04-02.log](https://github.com/zwave-js/node-zwave-js/files/14837689/zwavejs_2024-04-02.log)

**Comment by AlCalzone** [comment, marked as answer]
User codes are not automatically queried during the interview, since they are potentially a lot. You'll have to refresh them once after the interview.

**Comment by mohanblaze** [comment]
Hi @AlCalzone. Thanks for your immediate response. 

I wasn't expecting it to query automatically during the interview. I am trying to fetch only one user code at a time after the interview process, where it was giving a failed response. 

We noticed that, after the interview is done, there's no user code in slot 1, whereas the code 1111 works for the lock. We tried to fetch the user code using the mentioned API, but it gave us a failure response.

Thank you.

**Reply by AlCalzone** [reply]
> We tried to fetch the user code using the mentioned API, but it gave us a failure response.

That should not happen. Can you share logs of that? And what's the failure?

**Reply by mohanblaze** [reply]
Hi @AlCalzone.

Please find the attached log and screenshots. NodeId is 32.

You'll notice that in one of the screenshots with the error response. In that, I was hitting the userId status by enabling it. I was expecting to get the user code for user ID 1. At the same time, I confirm that the user code for userID 1 is 1111 and I was able to control the lock. Let me know, if you need any other details for a solution.

[zwavejs_2024-04-03.log](https://github.com/zwave-js/node-zwave-js/files/14848881/zwavejs_2024-04-03.log)
![Screenshot from 2024-04-03 13-27-45](https://github.com/zwave-js/node-zwave-js/assets/88714210/ce14ea8f-83fe-4ad4-8ca9-c4a4f53c1a06)
![Screenshot from 2024-04-03 13-26-39](https://github.com/zwave-js/node-zwave-js/assets/88714210/0c0fc9f3-6eb3-4016-bc2a-eec63680c93c)

**Reply by AlCalzone** [reply]
Oh, that's not querying the user code. You're basically setting the user code without specifying one.
@robertsLando there's no way in the UI to query individual user codes, is there?

The current value abstraction is a bit weird since status and user code belong together and have to be set at once. When setting a user code, the user ID status is automatically enabled, so you can just write into the user code field and it works. Only setting the user ID status doesn't work though.

**Reply by mohanblaze** [reply]
Thanks, @AlCalzone for the instant response. We were able to see the user code, once it was set. 

We need to retrieve the user code that has already been set. Is it possible to abstract the user code once? We require this feature and would appreciate your assistance with implementing it.

**Reply by AlCalzone** [reply]
You can read an individual user code using a [driver function](https://zwave-js.github.io/zwave-js-ui/#/usage/driver_function), which can also be done [via MQTT](https://zwave-js.github.io/zwave-js-ui/#/guide/mqtt?id=driverfunction):
```ts
const node = driver.controller.nodes.get(32); // <= node ID here
await node.commandClasses["User Code"].get(1); // <= user ID here
```

Querying individual user codes via the UI would be a feature request for [Z-Wave JS UI](https://github.com/zwave-js/zwave-js-ui)

**Reply by robertsLando** [reply]
> @robertsLando there's no way in the UI to query individual user codes, is there?

Nope

**Comment by danielsza** [comment]
While not exactly the answer that you’re looking for, keymaster https://github.com/FutureTense/keymaster might just do what you need. It displays user codes for each (enabled) slot…

**Reply by AlCalzone** [reply]
Or https://github.com/raman325/lock_code_manager for a solution that's tailored to Z-Wave locks
